File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Skills/SkillConfiguration.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.Bot.Builder.AI.Luis;
using Microsoft.Bot.Builder.AI.QnA;
using Microsoft.Bot.Builder.Azure;
using Microsoft.Bot.Configuration;
using Microsoft.Bot.Solutions.Middleware.Telemetry;

namespace Microsoft.Bot.Solutions.Skills
{
    public class SkillConfiguration : SkillConfigurationBase
    {
        public SkillConfiguration()
        {
        }

        public SkillConfiguration(BotConfiguration botConfiguration, Dictionary<string, Dictionary<string, string>> languageModels, string[] supportedProviders = null, string[] parameters = null, Dictionary<string, object> configuration = null)
        {
            if (supportedProviders != null && supportedProviders.Count() > 0)
            {
                IsAuthenticatedSkill = true;
            }

            foreach (var service in botConfiguration.Services)
            {
                switch (service.Type)
                {
                    case ServiceTypes.Generic:
                        {
                            if (service.Name == "Authentication")
                            {
                                var auth = service as GenericService;

                                foreach (var provider in supportedProviders)
                                {
                                    var matches = auth.Configuration.Where(x => x.Value == provider);

                                    foreach (var match in matches)
                                    {
                                        AuthenticationConnections.Add(match.Key, match.Value);
                                    }
                                }
                            }

                            break;
                        }
                }
            }

            foreach (var language in languageModels)
            {
                var localeConfig = new LocaleConfiguration
                {
                    Locale = language.Key
                };

                var path = language.Value["botFilePath"];
                var secret = language.Value["botFileSecret"];
                var config = BotConfiguration.Load(path, !string.IsNullOrEmpty(secret) ? secret : null);

                foreach (var service in config.Services)
                {
                    switch (service.Type)
                    {
                        case ServiceTypes.Dispatch:
                            {
                                var dispatch = service as DispatchService;
                                var dispatchApp = new LuisApplication(dispatch.AppId, dispatch.SubscriptionKey, dispatch.GetEndpoint());
                                localeConfig.DispatchRecognizer = new TelemetryLuisRecognizer(dispatchApp);
                                break;
                            }

                        case ServiceTypes.Luis:
                            {
                                var luis = service as LuisService;
                                var luisApp = new LuisApplication(luis.AppId, luis.SubscriptionKey, luis.GetEndpoint());
                                localeConfig.LuisServices.Add(service.Id, new TelemetryLuisRecognizer(luisApp));
                                break;
                            }

                        case ServiceTypes.QnA:
                            {
                                var qna = service as QnAMakerService;
                                var qnaEndpoint = new QnAMakerEndpoint()
                                {
                                    KnowledgeBaseId = qna.KbId,
                                    EndpointKey = qna.EndpointKey,
                                    Host = qna.Hostname,
                                };
                                var qnaMaker = new TelemetryQnAMaker(qnaEndpoint);
                                localeConfig.QnAServices.Add(qna.Id, qnaMaker);
                                break;
                            }
                    }
                }

                LocaleConfigurations.Add(language.Key, localeConfig);
            }

            if (parameters != null)
            {
                // add the parameters the skill needs
                foreach (var parameter in parameters)
                {
                    // Initialize each parameter to null. Needs to be set later by the bot.
                    Properties.Add(parameter, null);
                }
            }

            if (configuration != null)
            {
                // add the additional keys the skill needs
                foreach (var set in configuration)
                {
                    Properties.Add(set.Key, set.Value);
                }
            }
        }

        public override Dictionary<string, string> AuthenticationConnections { get; set; } = new Dictionary<string, string>();

        public override CosmosDbStorageOptions CosmosDbOptions { get; set; }

        public override Dictionary<string, LocaleConfiguration> LocaleConfigurations { get; set; } = new Dictionary<string, LocaleConfiguration>();

        public override Dictionary<string, object> Properties { get; set; } = new Dictionary<string, object>();
    }
}

File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/skills/skillConfiguration.ts
import { LuisApplication, QnAMakerEndpoint } from 'botbuilder-ai';
import { CosmosDbStorageSettings } from 'botbuilder-azure';
import { BotConfiguration, DispatchService, GenericService, IConnectedService,
    LuisService, QnaMakerService, ServiceTypes } from 'botframework-config';

import { ITelemetryLuisRecognizer, ITelemetryQnAMaker, TelemetryLuisRecognizer, TelemetryQnAMaker } from '../middleware/telemetry';
import { LocaleConfiguration } from './localeConfiguration';
import { SkillConfigurationBase } from './skillConfigurationBase';

export class SkillConfiguration extends SkillConfigurationBase {
    public authenticationConnections: { [key: string]: string } = {};
    public cosmosDbOptions!: CosmosDbStorageSettings;
    public localeConfigurations: Map<string, LocaleConfiguration> = new Map();
    public properties: { [key: string]: Object|undefined } = {};

    constructor(botConfiguration?: BotConfiguration,
                languagesModels?: Map<string, { botFilePath: string; botFileSecret: string }>,
                supportedProviders?: string[],
                parameters?: string[],
                configuration?: Map<string, Object>) {
        super();

        if (!botConfiguration || !languagesModels) {
            return;
        }

        if (supportedProviders && supportedProviders.length > 0) {
            this.isAuthenticatedSkill = true;
        }

        botConfiguration.services.forEach((service: IConnectedService) => {
            if (service.type === ServiceTypes.Generic && service.name === 'Authentication' && supportedProviders) {
                const auth: GenericService = <GenericService> service;

                supportedProviders.forEach((provider: string) => {
                    const matches: [string, string][] = Object.entries(auth.configuration)
                        .filter((val: [string, string]) => val[1] === provider);

                    matches.forEach((val: [string, string]) => {
                        this.authenticationConnections[val[0]] = val[1];
                    });
                });
            }
        });

        languagesModels.forEach((value: { botFilePath: string; botFileSecret: string }, key: string) => {
            const localeConfig: LocaleConfiguration = new LocaleConfiguration();
            localeConfig.locale = key;

            const config: BotConfiguration = BotConfiguration.loadSync(value.botFilePath, value.botFileSecret);

            config.services.forEach((service: IConnectedService) => {
                switch (service.type) {
                    case ServiceTypes.Dispatch: {
                        const dispatch: DispatchService = <DispatchService> service;

                        const dispatchApp: LuisApplication = {
                            applicationId: dispatch.appId,
                            endpointKey: dispatch.subscriptionKey,
                            endpoint: dispatch.getEndpoint()
                        };

                        localeConfig.dispatchRecognizer = new TelemetryLuisRecognizer(dispatchApp);
                        break;
                    }
                    case ServiceTypes.Luis: {
                        const luis: LuisService = <LuisService> service;

                        const luisApp: LuisApplication = {
                            applicationId: luis.appId,
                            endpointKey: luis.subscriptionKey,
                            endpoint: luis.getEndpoint()
                        };

                        const luisRecognizer: ITelemetryLuisRecognizer = new TelemetryLuisRecognizer(luisApp);
                        localeConfig.luisServices.set(luis.id, luisRecognizer);

                        break;
                    }
                    case ServiceTypes.QnA: {
                        const qna: QnaMakerService = <QnaMakerService> service;

                        const qnaEndpoint: QnAMakerEndpoint = {
                            knowledgeBaseId: qna.kbId,
                            endpointKey: qna.endpointKey,
                            host: qna.hostname
                        };

                        const qnaMaker: ITelemetryQnAMaker = new TelemetryQnAMaker(qnaEndpoint);
                        localeConfig.qnaServices.set(qna.id, qnaMaker);

                        break;
                    }
                    default:
                }
            });

            this.localeConfigurations.set(key, localeConfig);
        });

        if (parameters) {
            // Add the parameters the skill needs
            // Initialize each parameter to null. Needs to be set later by the bot.
            parameters.forEach((val: string) => this.properties[val] = {});
        }

        if (configuration) {
            // add the additional keys the skill needs
            configuration.forEach((val: Object, key: string) => this.properties[key] = val);
        }
    }
}
